src/workers/newNotificationV2Mail.ts

```diff
@@ -0,0 +1,702 @@
+
+    const latestPostRelation = await con.getRepository(PostRelation).findOne({
+      where: {
+        postId: post.id,
+        type: PostRelationType.Collection,
+      },
+      order: {
+        createdAt: 'DESC',
+      },
+      relations: {
+        relatedPost: {
+          source: true,
+        },
+      },
+    });
+
+    if (!latestPostRelation) {
+      return;
+    }
+
+    const latestRelatedPost =
+      (await latestPostRelation.relatedPost) as ArticlePost;
+    const latestRelatedPostSource = await latestRelatedPost.source;
+
+    return {
+      static: {
+        post_title: truncatePostToTweet(post),
+        post_image: post.image || pickImageUrl(post),
+        post_upvotes: post.upvotes,
+        post_comments: post.comments,
+        post_link: addNotificationEmailUtm(
+          notification.targetUrl,
+          notification.type,
+        ),
+        post_timestamp: formatMailDate(post.metadataChangedAt),
+        source_title: truncatePostToTweet(latestRelatedPost),
+        source_image:
+          latestRelatedPost.image || pickImageUrl(latestRelatedPost),
+        source_timestamp: formatMailDate(latestRelatedPost.createdAt),
+        source_name: latestRelatedPostSource.name,
+      },
+    };
+  },
+};
+
+const formatTemplateDate = <T extends TemplateData>(data: T): T => {
+  return Object.keys(data).reduce((acc, key) => {
+    if (typeof data[key] === 'number') {
+      return { ...acc, [key]: (data[key] as number).toLocaleString() };
+    }
+    return acc;
+  }, data);
+};
+
+const BATCH_SIZE = 100;
+const QUEUE_CONCURRENCY = 10;
+
+const worker: Worker = {
+  subscription: 'api.new-notification-mail',
+  handler: async (message, con, logger): Promise<void> => {
+    const data: Data = messageToJson(message);
+    const { id } = data.notification;
+    const [notification, attachments, avatars] =
+      await getNotificationV2AndChildren(con, id);
+    if (!notification) {
+      return;
+    }
+    const stream = await streamNotificationUsers(con, notification.id);
+    try {
+      await processStreamInBatches(
+        stream,
+        async (batch: { userId: string }[]) => {
+          const users = await con.getRepository(User).find({
+            where: {
+              id: In(batch.map((b) => b.userId)),
+              email: Not(IsNull()),
+              notificationEmail: notification.public ? true : undefined,
+            },
+          });
+          if (!users.length) {
+            return;
+          }
+          const templateData = await notificationToTemplateData[
+            notification.type
+          ](con, users, notification, attachments, avatars);
+          if (!templateData) {
+            return;
+          }
+          const formattedData = formatTemplateDate(templateData.static);
+          const personalized =
+            templateData.personalized?.map(formatTemplateDate);
+          const templateId = notificationToTemplateId[notification.type];
+          await sendEmail({
```

Comment:
not sure how relevant it is here, but do we get any info about which emails failed since we are processing these in batches? or do we even care about that here?

Reply:
Good point. See this: https://github.com/dailydotdev/daily-api/pull/1583#discussion_r1424015711